File: rewrite-xml/src/main/java/org/openrewrite/xml/AddToTagVisitor.java
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.xml;

import org.jspecify.annotations.Nullable;
import org.openrewrite.Cursor;
import org.openrewrite.Tree;
import org.openrewrite.marker.Markers;
import org.openrewrite.xml.tree.Content;
import org.openrewrite.xml.tree.Xml;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class AddToTagVisitor<P> extends XmlVisitor<P> {
    private final Xml.Tag scope;
    private final Xml.Tag tagToAdd;

    @Nullable
    private final Comparator<Content> tagComparator;

    public AddToTagVisitor(Xml.Tag scope, Xml.Tag tagToAdd) {
        this(scope, tagToAdd, null);
    }

    public AddToTagVisitor(Xml.Tag scope, Xml.Tag tagToAdd, @Nullable Comparator<Content> tagComparator) {
        this.scope = scope;
        this.tagToAdd = tagToAdd;
        this.tagComparator = tagComparator;
    }

    @Override
    public Xml visitTag(Xml.Tag t, P p) {
        if (scope.isScope(t)) {
            assert getCursor().getParent() != null;
            if (t.getClosing() == null) {
                t = t.withClosing(autoFormat(new Xml.Tag.Closing(Tree.randomId(), "\n",
                                Markers.EMPTY, t.getName(), ""), null, p, getCursor()))
                        .withBeforeTagDelimiterPrefix("");
            }

            //noinspection ConstantConditions
            if (!t.getClosing().getPrefix().contains("\n")) {
                t = t.withClosing(t.getClosing().withPrefix("\n"));
            }

            Xml.Tag formattedTagToAdd = tagToAdd;
            if (!formattedTagToAdd.getPrefix().contains("\n")) {
                formattedTagToAdd = formattedTagToAdd.withPrefix("\n");
            }
            formattedTagToAdd = autoFormat(formattedTagToAdd, null, p, getCursor());

            List<Content> content = t.getContent() == null ? new ArrayList<>() : new ArrayList<>(t.getContent());
            if (tagComparator != null) {
                int i = 0;
                for (; i < content.size(); i++) {
                    if (tagComparator.compare(content.get(i), formattedTagToAdd) > 0) {
                        content.add(i, formattedTagToAdd);
                        break;
                    }
                }
                if (i == content.size()) {
                    content.add(formattedTagToAdd);
                }
            } else {
                content.add(formattedTagToAdd);
            }

            t = t.withContent(content);
        }

        return super.visitTag(t, p);
    }

    /**
     * Add a tag to the children of another tag
     *
     * @param parent the tag that will have 'newChild' added to its children
     * @param newChild the tag to add as a child of 'parent'
     * @param parentCursor A cursor pointing one level above 'parent'. Determines the final indentation of 'newChild'.
     *
     * @return 'parent' with 'newChild' amongst its child elements
     */
    public static Xml.Tag addToTag(Xml.Tag parent, Xml.Tag newChild, Cursor parentCursor) {
        return addToTag(parent, parent, newChild, parentCursor);
    }

    /**
     * Add a tag to the children of another tag
     *
     * @param parentScope a tag which contains 'parent' as a direct or transitive child element.
     * @param parent the tag that will have 'newChild' added to its children
     * @param newChild the tag to add as a child of 'parent'
     * @param parentCursor A cursor pointing one level above 'parentScope'. Determines the final indentation of 'newChild'.
     *
     * @return 'parentScope' which somewhere contains 'parent' with 'newChild' amongst its child elements
     */
    public static Xml.Tag addToTag(Xml.Tag parentScope, Xml.Tag parent, Xml.Tag newChild, Cursor parentCursor) {
        //noinspection ConstantConditions
        return (Xml.Tag) new AddToTagVisitor<Void>(parent, newChild)
                .visitNonNull(parentScope, null, parentCursor);
    }
}

File: rewrite-csharp/csharp/OpenRewrite/Xml/AddToTagVisitor.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using OpenRewrite.Core;

namespace OpenRewrite.Xml;

/// <summary>
/// Adds a tag to the children of another tag.
/// Port of Java's org.openrewrite.xml.AddToTagVisitor.
/// </summary>
public class AddToTagVisitor<P>(Tag scope, Tag tagToAdd, Comparison<Content>? tagComparator = null) : XmlVisitor<P>
{
    public override Xml VisitTag(Tag tag, P p)
    {
        if (scope.Id == tag.Id)
        {
            var t = tag;

            // Ensure the tag has a closing element
            if (t.ClosingTag == null)
            {
                t = t.WithClosingTag(new Tag.Closing(
                        Guid.NewGuid(), "\n", Markers.Empty, t.Name, ""))
                    .WithBeforeTagDelimiterPrefix("");
            }

            // Ensure closing tag is on its own line
            if (!t.ClosingTag!.Prefix.Contains('\n'))
            {
                t = t.WithClosingTag(t.ClosingTag.WithPrefix("\n"));
            }

            var formattedTagToAdd = tagToAdd;
            if (!formattedTagToAdd.Prefix.Contains('\n'))
            {
                // Infer indentation from existing children or the closing tag
                var indent = InferChildIndent(t);
                formattedTagToAdd = formattedTagToAdd.WithPrefix("\n" + indent);
            }

            var content = t.ContentList == null ? new List<Content>() : new List<Content>(t.ContentList);
            if (tagComparator != null)
            {
                var i = 0;
                for (; i < content.Count; i++)
                {
                    if (tagComparator(content[i], formattedTagToAdd) > 0)
                    {
                        content.Insert(i, formattedTagToAdd);
                        break;
                    }
                }

                if (i == content.Count)
                {
                    content.Add(formattedTagToAdd);
                }
            }
            else
            {
                content.Add(formattedTagToAdd);
            }

            t = t.WithContentList(content);
            return base.VisitTag(t, p);
        }

        return base.VisitTag(tag, p);
    }

    /// <summary>
    /// Infer the indentation used by existing children of the tag.
    /// Falls back to the closing tag's prefix + 2 spaces, or just 4 spaces.
    /// </summary>
    private static string InferChildIndent(Tag tag)
    {
        // Try to get indent from existing children
        if (tag.ContentList != null)
        {
            foreach (var content in tag.ContentList)
            {
                if (content is Xml child)
                {
                    var prefix = child.Prefix;
                    var lastNewline = prefix.LastIndexOf('\n');
                    if (lastNewline >= 0)
                        return prefix[(lastNewline + 1)..];
                }
            }
        }

        // Infer from closing tag
        if (tag.ClosingTag != null)
        {
            var closingPrefix = tag.ClosingTag.Prefix;
            var lastNewline = closingPrefix.LastIndexOf('\n');
            if (lastNewline >= 0)
                return closingPrefix[(lastNewline + 1)..] + "  ";
        }

        return "    ";
    }
}
